File: server/controllers/companyController.js
const { ProfileRepository, HistoryRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Research company overview, interview stages, and technical expectations
// @route   POST /api/company/research
const researchCompany = async (req, res, next) => {
  try {
    let { company, role } = req.body;

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    if (profile) {
      if (!company) company = profile.company;
      if (!role) role = profile.role;
    }

    const targetCompany = company || 'Target Tech Company';
    const targetRole = role || 'Software Engineer';

    const prompt = `
Generate structured, professional company interview insights for "${targetCompany}" for the position of "${targetRole}".
Structure your research into:
1. "companyName": Name of the company
2. "overview": Concise summary of their market domain, engineering ethos, and core mission.
3. "majorProducts": Array of 3-4 key products, services, or platforms they are famous for.
4. "typicalInterviewStages": Array of standard stages (e.g. Online Assessment, Technical Round 1 - DSA, Technical Round 2 - System Design/Projects, HR/Values).
5. "coreTechnicalTopics": Array of 4-5 relevant tech areas commonly emphasized in their interview loops.
6. "preparationTips": 3-4 realistic actionable tips for preparing for this company.

Return ONLY valid JSON with this schema:
{
  "companyName": "...",
  "overview": "...",
  "majorProducts": ["...", "..."],
  "typicalInterviewStages": [
    { "stage": "...", "description": "..." }
  ],
  "coreTechnicalTopics": ["...", "..."],
  "preparationTips": ["...", "..."]
}
`;

    const context = { company: targetCompany, role: targetRole };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      context,
      jsonMode: true,
      moduleType: 'company-research',
      systemPrompt: 'You are an IT Industry Analyst and Tech Career Researcher.',
    });

    const data = aiResult.parsedJson || llmGateway.generateFallbackResponse('company-research', context);

    await HistoryRepository.create({
      userId: req.user._id,
      module: 'Company Research',
      company: targetCompany,
      role: targetRole,
      summary: `Researched interview stages and technical topics for ${targetCompany}`,
      metadata: { targetCompany },
    });

    res.status(200).json({
      success: true,
      data,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  researchCompany,
};

File: server/src/main/java/com/nexoffer/controller/CompanyController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/company")
public class CompanyController {

    private final ProfileRepository profileRepository;
    private final HistoryRepository historyRepository;
    private final LlmGatewayService llmGatewayService;

    public CompanyController(ProfileRepository profileRepository,
                             HistoryRepository historyRepository,
                             LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.historyRepository = historyRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping("/research")
    public ResponseEntity<?> researchCompany(@AuthenticationPrincipal User user, @RequestBody(required = false) Map<String, String> body) {
        String company = body != null ? body.get("company") : null;
        String role = body != null ? body.get("role") : null;

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        if (profileOpt.isPresent()) {
            Profile p = profileOpt.get();
            if (company == null || company.isBlank()) company = p.getCompany();
            if (role == null || role.isBlank()) role = p.getRole();
        }

        String targetCompany = (company != null && !company.isBlank()) ? company : "Target Tech Company";
        String targetRole = (role != null && !role.isBlank()) ? role : "Software Engineer";

        String prompt = "Generate structured, professional company interview insights for \"" + targetCompany + "\" for the position of \"" + targetRole + "\".\n" +
                "Structure your research into:\n" +
                "1. companyName\n" +
                "2. overview\n" +
                "3. majorProducts (array)\n" +
                "4. typicalInterviewStages (array of {stage, description})\n" +
                "5. coreTechnicalTopics (array)\n" +
                "6. preparationTips (array)\n\n" +
                "Return ONLY valid JSON with this schema:\n" +
                "{\n" +
                "  \"companyName\": \"...\",\n" +
                "  \"overview\": \"...\",\n" +
                "  \"majorProducts\": [\"...\"],\n" +
                "  \"typicalInterviewStages\": [{ \"stage\": \"...\", \"description\": \"...\" }],\n" +
                "  \"coreTechnicalTopics\": [\"...\"],\n" +
                "  \"preparationTips\": [\"...\"]\n" +
                "}";

        Map<String, String> context = new HashMap<>();
        context.put("company", targetCompany);
        context.put("role", targetRole);

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                "You are an IT Industry Analyst and Tech Career Researcher.",
                context,
                true,
                "company-research"
        );

        Object data = aiResponse.getParsedJson() != null ? aiResponse.getParsedJson() : llmGatewayService.generateFallbackResponse("company-research", context);

        historyRepository.save(new History(
                user.getId(),
                "Company Research",
                targetCompany,
                targetRole,
                "Researched interview rounds & tech focus areas"
        ));

        return ResponseEntity.ok(Map.of(
                "success", true,
                "data", data,
                "provider", aiResponse.getProvider()
        ));
    }
}
